### ISSUE TITLE
A version of compile that returns a side value

### ISSUE THREAD

**Issue by ChristopherKing42** [HeinrichApfelmus/reactive-banana#201]
I think it would be useful if there was a version of `compile` that in addition to compiling the network allowed `MomentIO` to return a side value:

> compile' :: MomentIO a -> IO (a, EventNetwork)

This allows you to, for example, extract the initial values of behaviors. Looking at the source code, it appears that there already is code to do the above, it just isn't exported in a nice way.

**Comment by ocharles** [comment]
I'd really like to have this, but am stuck in the age old problem of naming. I think changing `compile` would ultimately be the nicest, but it's just too big a breaking change to do without a deprecation cycle, which would itself the require another deprecation cycle to switch /back/ to `compile`. So I think it's best we just have both.

I don't like using a prime, as I associate that was strictness.

Any preferences for names? `compileAndReturn` maybe?

**Comment by ocharles** [comment]
Cc @HeinrichApfelmus and @mitchellwrosen

**Comment by ocharles** [comment]
The reason I'm interested in this is because it simplifies the following pattern:

```haskell
setup :: ... -> IO Callbacks
setup ... = do
  (aAddHandler, fireA) <- newAddHandler
  (bAddHandler, fireB) <- newAddHandler
  ...
  (zAddHandler, fireZ) <- newAddHandler

  actuate =<< compile do
    onA <- fromAddHandler
    onB <- fromAddHandler
    ...
    onZ <- fromAddHandler
    ...

  return Callbacks{ fireA, fireB, ..., fireZ }
```

to

```haskell
setup :: ... -> IO Callbacks
setup ... = do 
  (eventNetwork, callbacks) <- compile do
    (onA, fireA) <- newEvent
    (onB, fireB) <- newEvent
    ...
    (onZ, fireZ) <- newEvent
    ...
    return Callbacks{..}

  actuate eventNetwork
  return callbacks 
```

Smuggling that out with an `MVar` is certainly an option, but why jump through that hoop?

I'm also fairly liberal with breakage (and don't really think `reactive-banana` has a huge usage at the moment sadly), and would take a `compile` type change.

@HeinrichApfelmus Maybe you want to be a tie-breaker :smile:

**Comment by mitchellwrosen** [comment]
Good point, I like that pattern. Ok, I switch to (2), then (3), then (1).

**Comment by ChristopherKing42** [comment]
I wasn't sure how other people would feel about it, but if it's on the table option 2 is my personal favorite.

**Comment by ocharles** [comment]
Honestly, I think this is overly conservative. We could put network ids in `Event` and `Behavior` and check that at runtime, giving a good error message. My problem with this reasoning is it's not at all complete - I can still circumvent this by using an `IORef` or something, so life just becomes harder without actually _really_ solving anything. Personally I think the better documentation is to say that `compile` should never return `Event`s or `Behavior`s.

**Comment by ChristopherKing42** [comment]
> but eventually decided that this parameter made so much noise in the types that we would be better of with disallowing return values in compile

You could've just removed the "Frameworks t" constraint to get rid of most of the noise. The ST monad is the canonical example of using a type parameter to prevent leaks between different invocations; it doesn't use a type constraint.

**Comment by mitchellwrosen** [comment]
I agree with @ocharles, but I get where @HeinrichApfelmus is coming from. I think the pattern of "don't return this value here, which we do not attempt to prevent with the type system" is common (e.g. whenever acquiring a resource that's only usable before it's released/finalized).

And @ChristopherKing42, the "ST monad trick" is indeed how the old API used to work, and is precisely what "so much noise in the types" refers to - not the one `Frameworks t` constraint required at the call site of `compile`. And I'm sure this was not your intent, but I'd like to provide you with the feedback that beginning a comment with "you could've just <alternative>" in response to a justification for why a decision was made can come across as a bit rude, especially in textual form where one cannot discern your tone of voice ;)

**Comment by ChristopherKing42** [comment]
Right, that makes sense

> And I'm sure this was not your intent, but I'd like to provide you with the feedback that beginning a comment with "you could've just <alternative>" in response to a justification for why a decision was made can come across as a bit rude

Oh, sorry about that! Yeah, I was a bit rude in retrospect, sorry.

**Comment by mitchellwrosen** [comment]
Should this issue be closed? If I understand correctly, I'd summarize the discussion as:

- Ollie proposes allowing `compile` to return a value of any type, I (for what I'm worth) agree.
- @HeinrichApfelmus would like to keep the current API, since its type prevents the honest user from returning an `Event` or `Behavior` and passing along to another network.
